server/src/db.rs

```diff
@@ -115,6 +144,295 @@ impl Db {
+            for (partition_key, chunk_id) in partition_chunks {
+                self.drop_mutable_buffer_chunk(partition_key.as_str(), chunk_id)
+                    .await?;
+            }
+
+            self.capture_end_service_cycle("drop_chunks", now.elapsed());
+        }
+    }
+
+    /// Return all mutable buffer chunks of a given state
+    pub fn collect_chunks(&self, chunk_state: ChunkState) -> Result<Vec<(String, u32)>> {
+        // Get all partitions
+        let partition_keys = self.partition_keys().unwrap();
+
+        // Return a vector of (partition_key, chunk_id)
+        let mut partition_chunks: Vec<(String, u32)> = vec![];
+        for partition_key in partition_keys {
+            // Get all mutable buffer chunks of this partition_key of the given chunk_state
+            let chunks =
+                self.mutable_buffer_state_specified_chunks(partition_key.as_str(), chunk_state);
+            let mut part_chunks: Vec<(String, u32)> = chunks
+                .iter()
+                .map(|chunk| (partition_key.clone(), chunk.id()))
+                .collect();
+
+            partition_chunks.append(&mut part_chunks);
+        }
+
+        Ok(partition_chunks)
+    }
+
+    /// Get the mutable buffer chunk of a given partition key and chunk id
+    pub fn get_mutable_buffer_chunk(
+        &self,
+        partition_key: &str,
+        chunk_id: u32,
+    ) -> Result<Arc<Chunk>> {
+        self.mutable_buffer
+            .as_ref()
+            .context(DatatbaseNotWriteable)?
+            .get_chunk(partition_key, chunk_id)
+            .context(UnknownMutableBufferChunk { chunk_id })
+    }
+
+    /// Advance the state of a given chunk
+    pub fn advance_chunk_state(&self, partition_key: &str, chunk_id: u32) -> Result<()> {
+        // Get mutable buffer chunks of the (partition_key, chunk_id)
+        let mb_chunk = self.get_mutable_buffer_chunk(partition_key, chunk_id);
+
+        match mb_chunk {
+            Ok(chunk) => {
+                chunk.advance_state();
+                Ok(())
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Get all table names of a given chunk
+    pub fn get_chunk_tables(&self, partition_key: &str, chunk_id: u32) -> Result<Vec<String>> {
+        let mb_chunk = self.get_mutable_buffer_chunk(partition_key, chunk_id);
+
+        match mb_chunk {
+            Ok(chunk) => {
+                let tables = chunk
+                    .table_stats()
+                    .unwrap()
+                    .iter()
+                    .map(|stats| stats.name.clone())
+                    .collect();
+
+                Ok(tables)
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Check if all given tables are in the given chunk of the read buffer
+    pub fn all_tables_in_read_buffer(
+        &self,
+        partition_key: &str,
+        chunk_id: u32,
+        table_names: Vec<String>,
```

Comment:
Does this need to take ownership of the vector? Usually functions don't take ownership unless they need to, like to consume a value to transform it into something else and invalidate the passed-in value. I don't see any need here, so I'd write this as:

```suggestion
        table_names: &[String],
```

and then when this is called the argument will need to have a `&` in front of it.

Reply:
Right, no need ownership here